handlers.go

```diff
@@ -0,0 +1,59 @@
+package dagstore
+
+import (
+	"context"
+
+	"github.com/filecoin-project/dagstore/shard"
+)
+
+// RecoverImmediately takes a failureCh were DAGStore failures are sent, and
+// attempts to recover the shard immediately up until maxAttempts for each
+// unique shard.
+//
+// Attempt tracking does not survive restarts. When the passed context fires,
+// the failure handler goroutine will yield
+func RecoverImmediately(ctx context.Context, dagst *DAGStore, failureCh chan ShardResult, maxAttempts uint64) {
+	var (
+		recResCh = make(chan ShardResult, 128)
+		attempts = make(map[shard.Key]uint64)
+	)
+
+	for {
+		select {
+		case res := <-failureCh:
+			key := res.Key
+			att := attempts[key]
+			if att >= maxAttempts {
+				log.Infow("failure handler: max attempts exceeded; skipping recovery", "key", key, "from_error", res.Error, "attempt", att)
+				continue
+			}
+
+			log.Infow("failure handler: recovering shard", "key", key, "from_error", res.Error, "attempt", att)
+
+			// queue the recovery for this key.
+			if err := dagst.RecoverShard(ctx, key, recResCh, RecoverOpts{}); err != nil {
+				log.Warnw("failure handler: failed to queue shard recovery", "key", key, "error", err)
+				continue
```

Comment:
Here we bail out before incrementing the attempt count, so I think this will loop forever if RecoverShard always returns an error

Reply:
I perceive this behaviour as correct. Since our recovery request was **not accepted**, it cannot even be computed as "an attempt". This happens when a precondition is not met (e.g. shard does not exist), and we don't want to keep reattempting.

Reply:
This makes sense to me though something is really messed up if we're attempting to recover a Shard that does NOT exist.